Added ability to move users if streaming, doing nothing, and if only one activity being played.

stephenlee91 opened the pull request:
- Added ability to auto move users in voice channel if one or more people are streaming. If streaming VSCode the bot will determine if Study voice channel is already occupied. If occupied then the bot will move channel members to Lounge. Same for those who are streaming things other than VSCode.

- Added ability to auto move users in voice channel if noone is doing an activity. Everyone is moved to Lounge channel.

- Added ability to auto move users in voice channel if there is only one person playing a game and everyone else is not. Everyone will be moved to the voice channel of the game that one person is playing.

stephenlee91 commented:
Code has some slight problems that are very small but this is the main gist of what will be the movement. Any feedback would be great.

TimothyYLin commented on constantvariables.py:
```diff
@@ -6,4 +6,16 @@
+}
+
+DIFFERENT_NAME_CHANNELS = { # REQUIRED: Must have lounge, study, and cinema channels. You can name them whatever you want but make sure to replace the channel IDs with your server's
+    "Lounge": 1132567562014691418,
```
We're using a lot of channel IDs in our constant variables which is great. What is the best way to figure out the channel IDs of each server? Is there a script that we can run or put in our TODO so that people know how to do this? Can we make a wiki page for it as a HOW-TO?

stephenlee91 replied:
Channel ID has to be provided by the developer.

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+            channel: SequenceMatcher(None, channel_name.lower(), channel.name.lower()).ratio()
+            for channel in voice_channels
+        }
+        most_similar_channel = max(similarity_scores, key=similarity_scores.get)
```
Interesting way to write this. I haven't usually seen `...key=similarty_scores.get` as a way to find the max but rather something like `...key=lambda x: similarity_scores[x]` I think....

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+        most_similar_channel = max(similarity_scores, key=similarity_scores.get)
+        return most_similar_channel
+
+    async def move_users_to_channel(self, users, destination_channel):
```
Since these aren't be used as commands, we should "mark" these as private in python with `_` such as `_move_users_to_channel` and so that we're only using this in this cog (at least for now).

TimothyYLin replied:
A comment to mention that these are local helper functions will also be helpful.

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+    
+    def find_similar_channel(self, channel_name, voice_channels):
+        """Find a similar voice channel by name."""
+        similarity_scores = {
```
Out of curiosity, how do the similarty_scores look? IE: If I inputted "Rocket League" and we have "Rocket 🦵ue" in Konoha, what is the similarity score like?

stephenlee91 replied:
Rocket 🦵ue would be at the top of the list if there is no voice channel with the name "Rocket League". However I found some flaws with this function so I am still determining a better way to be able to find similar words

TimothyYLin replied:
Do you have any sample numbers? IE: The similarity score would be a 0.8 out of a scale of 0-1.

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+                if member.voice and member.voice.self_stream:
+                    activity_count["Streaming"] += 1
+
+                if activity_name in activity_count:  # Check if activity exists in the dictionary
```
Usage of a defaultdict might be helpful here

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+    async def automove(self, ctx):
+        voice_channels = ctx.guild.voice_channels
+        # Group users by their activities in voice channels
+        voice_channel_groups = {}
```
Comment of dictionary format might be helpful.

Structure for example voice_channel_groups = { "Valorant" : {"Streaming" : 1, "Nothing" : 2}, "Rocket 🦵ue" : {"Rocket League" : 3,  "Streaming" : 0, "Nothing" : 0} }

Or just a basic format writeup such as 
voice_channel_groups = { VC_Name1 : { Activity 1 : Count, Activity 2 : Count } , VC_Name2 : { Activity3 : Count, Activity 1 : Count } }

TimothyYLin commented on constantvariables.py:
```diff
@@ -6,4 +6,16 @@
+    "Rocket League": 3
+}
+
+DIFFERENT_NAME_CHANNELS = { # REQUIRED: Must have lounge, study, and cinema channels. You can name them whatever you want but make sure to replace the channel IDs with your server's
```
Can we make a ticket for not requiring these channels? This is ok for now but we'll need to allow for more flexibility in the future.

Have a check that if these channels do not exist in the constantvariables.py, then we just skip down the if/elif. We shouldn't have the bot break if servers have different configuration from ours. 

This could be something @asturrus could take a look at after the first iteration of this feature is merged. 
Please remember that we don't need to have the whole feature at perfection in first run especially since this is still an experimental bot. We can actually also open up a new branch for like a "beta" release and then cherry-pick the stable features from "beta" to the master branch.

stephenlee91 replied:
Must be required if developers want to use different names for their voice channels. For example, "Ichiraku" is known as the "Lounge" channel in our own server. Hard to know if "Ichiraku" is a lounge channel without context. Therefore it is necessary to require those channels to allow flexibility

TimothyYLin replied:
I completely agree with naming and that we can't figure out lounges or other channel types without context. But the "required" makes it seem like in order to use this bot, you must have lounge, study, and cinema channels. Which means in order to use our bot, your server needs to have a minimum of 3 voice channels. Some servers might want to only have 1-2 channels total and we can't be the ones that determine that. So we need flexibility so that the bot doesn't crash if we do a lookup for a "cinema" voice channel and it doesn't exist.

stephenlee91 replied:
As explained before, we need these different channels if we want to have usage of the bot. If we only have 1-2 channels then the bot does not serve much of a purpose in automove

TimothyYLin replied:
That's exactly the case. We're offering different features with the bot but it's not like we are turning on/off features. It's always good to have checks in case people do something they're not supposed to. 

Say there becomes a case where people take Rasenbot to mainly use it for the time tracking feature @benjaminl4917 is working on. The person who adds our bot takes a look at our Github wiki and sees the automove command. Without thinking they call the command and suddenly their bot crashes. There should never be a case where bot simply crashes. We need to keep the bot running through as many things as possible and output out error logs. 

If you want to make it required for people to have at least 3 channels to use Automove, that's fine. But then we should output an error log to the console for those that have less than 3 channels or don't have the correct specified channels and still attempt to call Automove.

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+        for channel in voice_channel_groups:  # Loop through voice channel groups
+            current_channel = discord.utils.get(ctx.guild.voice_channels, name=channel)
+
+            if voice_channel_groups[channel]["Streaming"] > 0: # if there is at least one person streaming
```
I suggest having ` >= 1` for readability since it's written out rather than having to think about "at least one person". Another method that works as being cleaner but not for readability is just to not have any comparison operators since 0's truthiness is `False` and anything else will be `True`.

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+
+            if voice_channel_groups[channel]["Streaming"] > 0: # if there is at least one person streaming
+                members_to_move = current_channel.members
+                if "Visual Studio Code" in voice_channel_groups[channel]:
```
Potential Failure Case: Someone is coding but a different person is streaming. IE: Person1 streaming RL but person2 is working on Visual Studio Code at the same time.

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+                    if self.bot.get_channel(DIFFERENT_NAME_CHANNELS["Study"]).voice_states:
+                        await self.move_users_to_channel_by_id(members_to_move, DIFFERENT_NAME_CHANNELS["Lounge"])
+                    else:
+                        await self.move_users_to_channel_by_id(members_to_move, DIFFERENT_NAME_CHANNELS["Study"])
```
Might be good to comment that Study is first choice but if full, then moved to Lounge.

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,95 @@
+                members_to_move = current_channel.members
+                await self.move_users_to_channel_by_id(members_to_move, DIFFERENT_NAME_CHANNELS["Lounge"])
+            
+            elif len(voice_channel_groups[channel]) == 3 and voice_channel_groups[channel][2] == 1: # If one person is playing a game and everyone else is not
```
Hard coded numbers are not a good idea here.

stephenlee91 replied:
Completely fine considering the structure of the dictionary here.

TimothyYLin replied:
Not necessarily. Dictionaries are not sorted and have no sense of ordering so we cannot guarantee that `voice_channel_groups[channel][2]` corresponds to some game and not `Nothing` or `Streaming`.

stephenlee91 replied:
So the keys and their corresponding values would be shifted around?

TimothyYLin replied:
Typically no but python makes no guarantees for dictionaries so yes there is a possibility that it can be shifted around.

LinsaneFC commented:
Curious... How long does it take the sorting to happen? We might be able to help test or make some more bots that can act as people in voice channels.

stephenlee91 commented:
Question: Do we still want to keep account if there are way too many people playing the same game? IE: 8 people playing Valorant. We discussed earlier that it is impossible to know who is in which team however it is still possible to move everyone to the same VC. Do we want to keep it to that level of simplicity or do we want to try something else?

stephenlee91 commented:
Question: If we want to move a group to a voice channel but that voice channel is occupied. Where do we want to move them? IE Group is going to move to Valorant but Valorant is already occupied (and there is no Valorant 2). Where should they go?

stephenlee91 commented:
Question: If there are multiple groups playing games (IE 3 people playing Valorant, 2 people playing Rocket League, 1 person playing League) how would we determine the movement? Where should the one person playing League move to?

stephenlee91 commented:
Question: Should we always consider if the destination channel will be occupied? IE Group needs to be moved to Valorant but Valorant is occupied. Might be redundant so may have to create a helper function that checks if a channel is occupied. If a channel is occupied where should they go?

LinsaneFC commented:
> Question: If we want to move a group to a voice channel but that voice channel is occupied. Where do we want to move them? IE Group is going to move to Valorant but Valorant is already occupied (and there is no Valorant 2). Where should they go? Should we have a general channel?

I'm ok with 2 scenarios that both start with...
Prereq: Check the Valorant channel that is full and determine that the people there are actually already playing Valorant. If they are not, move them as well first. 

Then... If the Valorant channel is full and people belong in Valorant there...
1. Don't move this current group.
2. Move them to a general channel.

LinsaneFC commented:
> Question: If there are multiple groups playing games (IE 3 people playing Valorant, 2 people playing Rocket League, 1 person playing League) how would we determine the movement? Where should the one person playing League move to?

Depends on their starting position. 
If they are all in separate voice channels to begin with, move them into their own gaming channels (IE: 3 people to Valorant, 2 people to Rocket League, 1 person by themselves in League channel). 
If they are all in the same channel and don't want to split up, then I think our previous discussions would put them in Valorant (where the majority are).

Since we are doing this by commands, perhaps we can do a command to "force" movement. 
IE: The people are all playing different games and are in the same channel to start. They know they want to split it. Current priority keeps them together. So have an option to force the split and they'll all go to their respective channels.

LinsaneFC commented:
> Question: Should we always consider if the destination channel will be occupied? IE Group needs to be moved to Valorant but Valorant is occupied. Might be redundant so may have to create a helper function that checks if a channel is occupied. If a channel is occupied where should they go?

This question is sort of answered here: https://github.com/LinsaneFC/RasenBot/pull/23#issuecomment-1668488567. Check if the channel is occupied but also ensure the occupants belong there in the first place.

stephenlee91 commented:
- Added in a flag that checks if the auto move ability is activity
- Completed streaming, noone, and one activity being played movement
- Also added in a check for if a destination channel is occupied. If occupied then checks if there is a second channel for same activity. If not then moved to general channel
- Changed voice_channel_groups structure to include members object in a list (voice_channel_groups would look like this: { VC_Name1 : { Activity 1 : [], Activity 2 : [] } , VC_Name2 : { Activity3 : [], Activity 1 : [] } })
- _find_similar_channel confirmed to be working for the necessary threshold and similar channel names that would be expected

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -1,13 +1,115 @@
+                current_channel = discord.utils.get(ctx.guild.voice_channels, name=channel)
+                members_to_move = current_channel.members
+
+                if len(voice_channel_groups[channel]["Streaming"]) >= 1: # if there is at least one person streaming
```
What happens if the person is streaming Valorant, RL, or some other game? We still go to Lounge or Study?

TimothyYLin replied:
There should be a way to differentiate streaming brave, chrome, vlc player, etc. vs a game that exists.

TimothyYLin commented on constantvariables.py:
```diff
@@ -6,4 +6,13 @@
\ No newline at end of file
+WELCOME_REACT_EMOJI = "👍" # Replace with the emoji you want to use for your welcome message react emoji
+
+AUTO_MOVE_FLAG = True # REQUIRED: If you want to use the auto move ability for RasenBot then set to True. If not then set to False
```
Good idea. In the future, let's have commands that can write into this file as "settings" so that people can type commands to the bot to turn on/off features.

stephenlee91 reviewed (commented):
Finished requested changes from @LinsaneFC about user streaming an activity

TimothyYLin commented on cogs/moveusers.py:
```diff
@@ -80,6 +84,8 @@ async def automove(self, ctx):
                                 await self._move_users_to_channel_by_id(members_to_move, DIFFERENT_NAME_CHANNELS["Lounge"])
                             else:
                                 await self._move_users_to_channel_by_id(members_to_move, DIFFERENT_NAME_CHANNELS["Study"])
+                        elif voice_channel_groups[channel]["Streaming"][member_screen_shares] != "General":
```
Not the best to do a not operator here. If possible just always stick with positives for easier understanding. Especially since it's a continue, let's do a continue for else and check for general and if so move to lounge or cinema

stephenlee91 committed:
Added ability to automove users if there are users streaming, doing nothing, or if only one activity is being played.
Added functionality to turn on/off automove based on a flag added to constantvariables.py
Improved channel string matching threshold, checks for user streaming an activity, and overall readability based on reviewer comments.
